fix pod lib lint errors for pod with test_spec

paulz opened the pull request:
1. Generate info plist file for test spec target to create valid test bundle, allowing tests to load bundle resources fixes https://github.com/CocoaPods/CocoaPods/issues/7506
2. Clear product module name to inherit Xcode default value, which is `$(PRODUCT_NAME:c99extidentifier)`, see https://github.com/CocoaPods/CocoaPods/pull/7510#discussion_r174983832. This always creates valid identifier different from the parent spec module name. 

closes https://github.com/CocoaPods/CocoaPods/issues/7506

CocoaPodsBarista commented:
<!--
  1 Error: Please include a CHANGELOG ent...
  0 Warnings
  0 Messages
  1 Markdown
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry to credit yourself! <br />
You can find it at <a href="https://github.com/CocoaPods/CocoaPods/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

Here's an example of your CHANGELOG entry:

```markdown
* fix pod lib lint errors for pod with test_spec  
  [paulz](https://github.com/paulz)
  [#issue_number](https://github.com/CocoaPods/CocoaPods/issues/issue_number)
```

*note*: There are two invisible spaces after the entry's text.


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

paulz commented:
@dnkoutso would you review?

orta commented on CHANGELOG.md:
```diff
@@ -780,7 +784,7 @@ To install release candidates run `[sudo] gem install cocoapods --pre`
+* Move Installer target verification into the Xcode namespace
   [Danielle Tomlinson](https://github.com/DanToml)
   [#5607](https://github.com/CocoaPods/CocoaPods/pull/5607)
 
```
Any chance you can undo these whitespace changes and re-add your entry with format on save disabled? (if you're using vscode that's `cmd + alt + shift + s` )

paulz replied:
re-pushed the clean changes, thank you. will check out the vscode.

dnkoutso commented on lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb:
```diff
@@ -277,7 +277,7 @@ def add_test_targets
                 # irrelevant to whether we use frameworks or not.
                 configuration.build_settings['PRODUCT_NAME'] = name
-                configuration.build_settings['PRODUCT_MODULE_NAME'] = name
+                configuration.build_settings.delete('PRODUCT_MODULE_NAME')
```
what is the default value Xcode produces in this case?

dnkoutso commented on CHANGELOG.md:
```diff
@@ -55,6 +55,10 @@ To install release candidates run `[sudo] gem install cocoapods --pre`
 
 ##### Bug Fixes
 
+* fix pod lib lint errors for pod with test_spec
```
please be more descriptive on this change. "Generate `Info.plist` for test targets"` for example.

Also update your commit message to reflect this similarly.

paulz replied:
yes, great idea, added more details.

paulz commented:
thank you for the review, learning a lot in the process!

segiddins commented on lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb:
```diff
@@ -277,7 +277,10 @@ def add_test_targets
+                # Use xcode default product module name, which is $(PRODUCT_NAME:c99extidentifier)
+                # this gives us always valid name that is distinct from the parent spec module name
+                # which allow tests to use either import or @testable import to access the parent framework
+                configuration.build_settings.delete('PRODUCT_MODULE_NAME')
```
I dont think this will work, since we override `PRODUCT_MODULE_NAME` elsewhere

dnkoutso replied:
we do but that happens earlier than this.

paulz replied:
looks like you right, as value gets inherited from parent spec and breaks compilation:
https://travis-ci.org/paulz/ImageCoordinateSpace/builds/354136985#L2284
```
The following build commands failed:
	CompileSwiftSources normal i386 com.apple.xcode.tools.swift.compiler
(1 failure)
 -> ImageCoordinateSpace (1.0.5)
    - ERROR | [iOS] xcodebuild: Returned an unsuccessful exit code.
    - NOTE  | [iOS] xcodebuild:  <unknown>:0: error: module name "ImageCoordinateSpace-Unit-Tests" is not a valid identifier
```

when cleared the default Xcode value works, here is branch demonstrating the successful `pod lib lint`: https://github.com/paulz/ImageCoordinateSpace/tree/test-spec on travis: https://travis-ci.org/paulz/ImageCoordinateSpace/builds/354903900

dnkoutso commented on lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb:
```diff
@@ -291,6 +294,11 @@ def add_test_targets
 
+              # Generate vanila Info.plist for test target similar to the one xcode gererates for new test target.
+              # This creates valid test bundle accessible at the runtime, allowing tests to load bundle resouces.
+              test_info_plist_path = project.path.dirname + "#{name}/#{test_type.capitalize}TestInfo.plist"
```
this is a good fix, does it happen for normal targets within their create target method or is it a separate method?

dnkoutso replied:
This path seems wrong 
![screen shot 2018-03-19 at 5 33 40 pm](https://user-images.githubusercontent.com/310370/37629372-b6277ff6-2b9b-11e8-8454-3bc37f8b26eb.png)

The file reference is actually OK because its a different path but it should go under `Target Support Files` to be organized like the rest of them.

paulz replied:
got it

dnkoutso commented on CHANGELOG.md:
```diff
@@ -55,6 +55,10 @@ To install release candidates run `[sudo] gem install cocoapods --pre`
 
 ##### Bug Fixes
 
+* Fix `test_spec` target generation to pass `pod lib lint`  
```
@paulz this changelog entry still does not make sense or convey enough information on what the change actually was.

I would split this into two commits and two changelog entries that includes:

1) Create a generic Info.plist file for test targets
2) Use `PRODUCT_MODULE_NAME` for generated test targets

paulz replied:
will do

dnkoutso commented:
@paulz please rebase and squash into tow single commits, one that includes the fix for `PRODUCT_MODULE_NAME` and one for `Info.plist` generation.

Also please explain why `Info.plist` is needed again, it says about loading resources but do you mean for resources directly bundled with the test bundle or via a separate resource bundle target?

dnkoutso commented:
Can you provide a specific example in which the lack of an info plist for a test target causes a problem?

paulz commented:
@dnkoutso for example this code fails to run when there is no info plist:
```
            let testBundle = Bundle(for: type(of: self))
            let image = UIImage(named: "rose", in: testBundle, compatibleWith: nil)!
```
the returned image is just nil and when you open the project in Xcode it shows gray screen on Info tab in target settings.

dnkoutso commented:
@paulz and I assume this means you are using `test_spec.resource` in your podspec for the test spec? Can you paste it? 

Sorry for the scrutiny just want to double check we are fixing the right thing :)

paulz commented:
no problem, thank you for being so responsive! new to this code, and very grateful for cocoapods, as I've been using it for years.

Here is the example with resource:
https://github.com/paulz/ImageCoordinateSpace/blob/test-spec/ImageCoordinateSpace.podspec#L23

was failing here:
https://github.com/paulz/ImageCoordinateSpace/blob/test-spec/Unit%20Tests/ConvertFromCoordinateSpaceSpec.swift#L15

dnkoutso commented:
@paulz perfect this was exactly what I was looking for.

paulz commented:
great, thank you, squashed into 2 commits, let me know if there are more changes

dnkoutso commented on lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb:
```diff
@@ -291,6 +294,12 @@ def add_test_targets
+              # Generate vanila Info.plist for test target similar to the one xcode gererates for new test target.
+              # This creates valid test bundle accessible at the runtime, allowing tests to load bundle resources
+              # defined in podspec.
+              test_info_plist_path = target.support_files_dir + "#{test_type.capitalize}TestInfo.plist"
```
very nit but move this into a method in `pod_target.rb`

```ruby
    # @param  [Symbol] test_type
    #         The test type this Info.plist path is for.
    #
    # @return [Pathname] The absolute path of the Info.plist for the given test type.
    #    
    def info_plist_path_for_test_type(test_type)
      support_files_dir + "#{test_target_label(test_type)}-Info.plist"
    end
```

Then call it from here. We will soon be refactoring `pod_target.rb` but for now it's OK to add this there.

paulz replied:
yes, looks like opportunity to refactor, will do

dnkoutso commented:
Minor nit for extracting method out and I think this is great!

paulz commented:
extracted `target.info_plist_path_for_test_type(test_type)` method and added spec for it. Note the expected path now looks like `CoconutLib-Unit-Tests-Info.plist` with dashes between words.

dnkoutso commented:
I thnk thats OK and shouldnt break anything

dnkoutso commented:
LGTM! Will merge on green

dnkoutso commented:
@paulz thanks! Please also try it locally if you haven't already on your issue and ensure the test spec works!
